### PR TITLE
Prototype of much of the PEP

### PR DESCRIPTION
**Pull request by JelleZijlstra**
This isn't fully functional, but it should be enough to unblock work
on the runtime. I'll update this PR as I implement more.

### CODE REVIEW

**Comment by gvanrossum** [comment]
If you merge Eric's updated tests, do they pass? What's still missing?

**Comment by JelleZijlstra** [comment]
I haven't looked at his tests yet (might spend some time on that tonight), but I don't think a lot of PEP 695-specific syntax will work until we get symtable changes implemented.

I was testing by just running `test_typing.py` (this branch replaces `typing.TypeVar` with the C version). I think there's ~140 failing tests; a lot around `*TypeVarTuple` not working, and a lot dealing with various aspects of TypeVar substitution.

**Comment by gvanrossum** [comment]
Maybe we ought to focus on the symtable work first then? (In my experience the syntax *works* but has no semantics, but maybe I didn't explore deeply enough.)

**Comment by JelleZijlstra** [comment]
I got almost all tests in test_typing.py to pass by calling back into typing.py from the C implementation; the remaining ones are a little tedious but should be doable too. I'll start looking into the symtable now.

**Comment by cdce8p** [comment]
@JelleZijlstra FYI I've created #7 with the AST changes I suggested [here](https://discuss.python.org/t/pep-695-type-parameter-syntax/21646/99).

File: Objects/typevarobject.c
Comment (Eclips4) [inline Objects/typevarobject.c:368]:
```suggestion
    }
    else if (n_constraints == 0) {
        constraints = NULL;
    } 
    else if (bound != NULL) {
        PyErr_SetString(PyExc_TypeError,
                        "Constraints cannot be combined with bound=...");
        Py_XDECREF(bound);
        return NULL;
    }
```
PEP 7

**Comment by sunmy2019** [comment]
[6312d7b](https://github.com/erictraut/cpython/pull/3/commits/6312d7bdd84f6821a307b4a09156d885a6230f2d) got test failure (segment fault) for `test_compile`.

By the way, why is this PR so huge? Shouldn't it be a series of small PRs that merged to [erictraut:type_param_syntax4](https://github.com/erictraut/cpython/tree/type_param_syntax4) somehow?

**Comment by sunmy2019** [comment]
> [6312d7b](https://github.com/erictraut/cpython/pull/3/commits/6312d7bdd84f6821a307b4a09156d885a6230f2d) got test failure (segment fault) for `test_compile`.

Minimal case
```py
import ast

_ast = compile("class _: pass", "2", "exec", ast.PyCF_ONLY_AST)
compile(_ast, "", "exec")
```

File: Python/compile.c
Comment (sunmy2019) [inline Python/compile.c:2089]: The newest crash happens if `n==0`. Not guarded here.

I guess 0 should not be allowed here, somewhere else is broken. Also, an assert should be added here.

**Review by sunmy2019 — CHANGES_REQUESTED** [review]

File: Python/compile.c
Comment (JelleZijlstra) [reply Python/compile.c:2089]: Thanks, I thought it would be NULL if there are no type params but apparently that's not true. I pushed a fix.

**Comment by JelleZijlstra** [comment]
> [6312d7b](https://github.com/erictraut/cpython/pull/3/commits/6312d7bdd84f6821a307b4a09156d885a6230f2d) got test failure (segment fault) for `test_compile`.
> 
> By the way, why is this PR so huge? Shouldn't it be a series of small PRs that merged to [erictraut:type_param_syntax4](https://github.com/erictraut/cpython/tree/type_param_syntax4) somehow?

I can't merge into Eric's branch, and by now there's enough for a reasonably complete CPython PR. Unfortunately it's hard to split up the change as there are few self-contained subparts.

**Comment by erictraut** [comment]
@JelleZijlstra, would you like me to merge into my branch now? I could also create a new branch (e.g. `erictraut:type_param_syntax5`) that rebases all of these changes onto the latest cpython main branch. That would make it easier to tell what you changed.

**Comment by erictraut** [comment]
Or would you prefer to submit the PR to cpython directly from your branch?

**Comment by JelleZijlstra** [comment]
I already submitted a draft PR with my branch, I think that's a good outcome. Your parser changes seem solid, I haven't needed to change anything there.

I talked to Thomas about the change I suggested in #11 and he's receptive to it, but we will need SC approval. I can handle the communication for that. I'm working with various other core devs too to get advice.

**Comment by erictraut** [comment]
Sounds good. Thanks for all of your work on this! Looks like you've made great progress.

**Comment by sunmy2019** [comment]
I will keep working on reviewing this in the next couple of weeks.

I will move to that PR.

**Comment by gvanrossum** [comment]
I propose that Eric close this in favor of https://github.com/python/cpython/pull/103764

**Comment by erictraut** [comment]
Closed.

**Pull request closed without merging**
